software/firmware/europi.py

```diff
@@ -298,16 +298,17 @@ def _bounce_wrapper(self, pin):
                 return
             self.last_rising_ms = time.ticks_ms()
             return self._rising_handler()
-        elif self.value() == LOW:
+        else:
             if time.ticks_diff(time.ticks_ms(), self.last_falling_ms) < self.debounce_delay:
                 return
             self.last_falling_ms = time.ticks_ms()
 
-            # Check if 'other' pin is set and if 'other' pins is high and if this pin has been high for long enough.
+            # When this button is released after a short press, check if 'other' button is set and currently pressed not long ago.
             if (
                 self._other
                 and self._other.value()
-                and time.ticks_diff(self.last_falling_ms, self.last_rising_ms) > 500
+                and time.ticks_diff(self.last_falling_ms, self.last_rising_ms) > 200
+                and time.ticks_diff(self.last_falling_ms, self._other.last_rising_ms) < 1000

```

Comment:
[question] I'm having trouble understanding the logic changes to the 'both' functionality. Could you more explicitly explain what is wrong with the current logic and how this change addresses that?

Reply:
> I was always thinking: Pressing button 1 does this, pressing button 2 does that and pressing both together does something else. But the old implementation is more like have both buttons down for some time.

Some history: the 'both' functionality was added to enable a global action that would quit a script and return to the menu. We wanted something that was simple to trigger, but wouldn't happen accidentally. Technically, scripts could use 'both' for their own features, but then they couldn't participate in the menu. So practically, the 'both' feature is only used to quit back to the menu. Originally the action was 'hold one button and tap the other' but this proved awkward to perform and explain. It was changed to 'hold both buttons for 500ms' to satisfy the 'simple but not accidental' requirement. So 'have both buttons down for some time' is an accurate description.

>It didn't check if the first button was maybe pressed down 10s ago and just happens to be still down (triggers when it shouldn't based on my opinion).

I'm not sure I understand the situation where this case would occur, and the user wasn't trying to trigger the 'both' function. Do you have an example?

>Also if both buttons where pressed and released faster than 500ms it wouldn't trigger (doesn't trigger where it should based on my opinion).

This is intentional, to avoid accidental triggering of the 'both' function. This might occur if the user is navigating the script's interface quickly, utilizing both buttons. By requiring that the user hold for half a second we can be more reasonably sure of their intention.

>However I am fine with removing this part of the PR and maybe alter the documentation to more clearly state what conditions are needed to trigger the 'both' action

I'm always in favor of clarifying documentation. As for removing this change, I'm leaning towards the existing, simpler implementation, but then again I wrote it and I may be biased. I also haven't heard any complaints about the current operation (maybe I've missed them). I'd prefer if changes that affected a core functionality like this were made based on the input of many users.